plugins/templates/create.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const boom = require('boom');
+const schema = require('screwdriver-data-schema');
+const hoek = require('hoek');
+const urlLib = require('url');
+
+module.exports = () => ({
+    method: 'POST',
+    path: '/templates',
+    config: {
+        description: 'Create a new template',
+        notes: 'Create a specific template',
+        tags: ['api', 'templates'],
+        auth: {
+            strategies: ['token', 'session'],
+            scope: ['build']
+        },
+        plugins: {
+            'hapi-swagger': {
+                security: [{ token: [] }]
+            }
+        },
+        handler: (request, reply) => {
+            const pipelineFactory = request.server.app.pipelineFactory;
+            const templateFactory = request.server.app.templateFactory;
+            const token = request.auth.credentials;
+            const pipelineId = token.pipelineId;
+            const name = request.payload.name;
+            const version = request.payload.version;
+            const labels = request.payload.labels || [];
+
+            return Promise.all([
+                pipelineFactory.get(pipelineId),
+                templateFactory.list({ name })
+            ]).then(([pipeline, templates]) => {
+                const templateConfig = hoek.applyToDefaults(request.payload, {
+                    scmUri: pipeline.scmUri,
+                    labels
+                });
+
+                // If template doesn't exist yet, just create a new entry
+                if (templates.length === 0) {
+                    return templateFactory.create(templateConfig);
+                }
+
+                // If template exists, but this build's scmUri is not the same as template's scmUri
+                // Then this build does not have permission to publish
+                if (pipeline.scmUri !== templates[0].scmUri) {
+                    throw boom.unauthorized('Not allowed to publish this template');
+                }
+
+                // If template exists and has good permission, check the exact version
+                return templateFactory.get({ name, version })
+                    .then((template) => {
+                        // If the version doesn't exist, create a new entry
```

Comment:
This endpoint is more for `CREATE`, so that's why I made it a `POST`. 
The only time it does an update is when the exact same name and version already exist. And I added that guard because of what we discussed here: https://github.com/screwdriver-cd/screwdriver/issues/470#issuecomment-283182395 on how to publish a template. We would always call a `POST /templates` in the yaml. It's a little confusing to do an update inside a create, but I think `POST` is more appropriate since it's supposed to create, not update. And the same version is a special case. Also, we don't have `id` so we can't do`PUT /templates/{id}`.